contracts/access/manager/AccessManager.sol

```diff
@@ -0,0 +1,703 @@
+ * These restrictions are expressed in terms of "groups".
+ *
+ * An AccessManager instance will define a set of groups. Accounts can be added into any number of these groups. Each of
+ * them defines a role, and may confer access to some of the restricted functions in the system, as configured by admins
+ * through the use of {setFunctionAllowedGroup}.
+ *
+ * Note that a function in a target contract may become permissioned in this way only when: 1) said contract is
+ * {AccessManaged} and is connected to this contract as its manager, and 2) said function is decorated with the
+ * `restricted` modifier.
+ *
+ * There is a special group defined by default named "public" which all accounts automatically have.
+ *
+ * Contracts where functions are mapped to groups are said to be in a "custom" mode, but contracts can also be
+ * configured in two special modes: 1) the "open" mode, where all functions are allowed to the "public" group, and 2)
+ * the "closed" mode, where no function is allowed to any group.
+ *
+ * Since all the permissions of the managed system can be modified by the admins of this instance, it is expected that
+ * they will be highly secured (e.g., a multisig or a well-configured DAO).
+ *
+ * NOTE: This contract implements a form of the {IAuthority} interface, but {canCall} has additional return data so it
+ * doesn't inherit `IAuthority`. It is however compatible with the `IAuthority` interface since the first 32 bytes of
+ * the return data are a boolean as expected by that interface.
+ */
+contract AccessManager is Context, Multicall, IAccessManager {
+    using Time for *;
+
+    uint256 public constant ADMIN_GROUP = type(uint256).min; // 0
+    uint256 public constant PUBLIC_GROUP = type(uint256).max; // 2**256-1
+
+    mapping(address target => AccessMode mode) private _contractMode;
+    mapping(address target => mapping(bytes4 selector => uint256 groupId)) private _allowedGroups;
+    mapping(uint256 groupId => Group) private _groups;
+    mapping(bytes32 operationId => uint48 schedule) private _schedules;
+    mapping(bytes4 selector => Time.Delay delay) private _adminDelays;
+
+    // This should be transcient storage when supported by the EVM.
+    bytes32 private _relayIdentifier;
+
+    /**
+     * @dev Check that the caller has a given permission level (`groupId`). Note that this does NOT consider execution
+     * delays that may be associated to that group.
+     */
+    modifier onlyGroup(uint256 groupId) {
+        address msgsender = _msgSender();
+        if (!hasGroup(groupId, msgsender)) {
+            revert AccessControlUnauthorizedAccount(msgsender, groupId);
+        }
+        _;
+    }
+
+    /**
+     * @dev Check that the caller is an admin and that the top-level function currently executing has been scheduled
+     * sufficiently ahead of time, if necessary according to admin delays.
```

Comment:
What about mentioning this as well?

```suggestion
     * sufficiently ahead of time, if necessary according to admin delays.
     *
     * IMPORTANT: This modifier relies on `msg.sig`. It'll share permissions with the `0x00000000`
     * selector if used in the `receive` or `fallback` function.
```

Reply:
There are two things:
- the _adminDelay is shared by msg.sig, so if we put a delay on function `0x00000000` it will also apply to the receive function (if for any strange reason someone adds that modifier to receive or fallback).
- the canCall group/permission system does not distinguish between function `0x00000000` and `receive`. This is affecting all `restricted()` function, not just the onlyAdmin here.

The first point should be documented in `setAdminFunctionDelay`

The second point should be documented somewhere where its visible. Possibly in both `setFunctionAllowedGroup` and `AccessManaged`

Reply:
I agree with both, let's document them. However, I am not sure about the locations. My proposal would be to mention in the contract's NatSpec that any operation involving a selector may consider `0x00000000` collisions. This way it's more succinct than adding the note in multiple places.

Currently, a similar note is in the `restricted` function but not in `canCall`. I'd be in favor of your suggestions, but just don't want to bloat the documentation with this warning everywhere

Reply:
There is already a note about the receive function in AccessManaged.

For AccessManager, we should detect if msg.data.length < 4.